docs/proposals/templates.md

```diff
@@ -0,0 +1,349 @@
+    of allowing for full complex-type substitution.
+* Parameter metadata.  Parameters should include at a minimum, information describing the purpose of the parameter, whether it is 
+  required/optional, and a default/suggested value.  Type information could also be required to resolve the non-string substitution
+  requirement.
+* Template metadata.  Templates should be able to include metadata describing their purpose or links to further documentation and 
+  versioning information.
+* Pluggable template expansion.  It should be possible to define multiple template expansion strategies and the system should
+  delegate a template to the correct plugin for expansion and then return the result.
+
+## Proposed Object Syntax
+
+Example of a template definition:
+```
+kind: TemplateDeclaration
+apiVersion: v1
+metadata: 
+  name: myHelloWorldTemplate
+templateKind: python
+params:
+  - name: message
+    required: true
+    default_value: Hello World!
+    type: string
+  - name: port
+    required: true
+    default_value: 8080
+    type: int
+includes:
+  - path: https://github.com/myorg/mytemplates/template1.yml
+  - path: https://github.com/myorg/mytemplates/template2.yml
+body: |
+  import json
+  import kubernetes_tmpl
+  # Real templates would be doing actual computation.
+  print json.dumps({
+    'apiVersion': 'v1',
+    'kind': 'Pod',
+    'metadata': {
+      'name': kubernetes_tmpl.GetVar('message'),
+      'labels': {'name': 'nginx'},
+    },
+  })
+```
+
+Where "templateBody" is arbitrary content that will be processed by the appropriate template plugin, as selected by templateKind.
+
+Because the template body is arbitrary and various plugins can be provided, the overall capability and syntax of templates 
+is not dictated by this specification.
+
+The schema also provides for an array of references to additional templates that should be expanded when this template is
+expanded.
+
+Additional example ("openshift" syntax):
+```
+kind: "TemplateDeclaration"
+apiVersion: "v1"
+metadata: 
+  name: "mongodb-ephemeral"
+  annotations: 
+    description: "Provides a MongoDB database service"
+templateKind: openshift
+parameters: 
+  - 
+    name: "DATABASE_SERVICE_NAME"
+    description: "Database service name"
+    value: "mongodb"
+    required: true
+    type: "string"
+  - 
+    name: "MONGODB_USER"
+    description: "Username for MongoDB user that will be used for accessing the database"
+    required: true
+    type: "string"
+  - 
+    name: "MONGODB_PASSWORD"
+    description: "Password for the MongoDB user"
+    required: true
+    type: "string"
+  - 
+    name: "MONGODB_DATABASE"
+    description: "Database name"
+    value: "sampledb"
+    required: true
+    type: "string"
+  - 
+    name: "REPLICA_COUNT"
+    description: "Number of mongo replicas to run"
+    value: "1"
+    required: true
+    type: "int"
+body: |
```

Comment:
Note that it's ok if the replica count field contains a string instead of an int, as we aren't validating the data at this point.  The only requirement is that it parses as YAML.